Enhance RawPOSIX Dispatcher with Memory Validation and Improved Syscall Handling in Nacl

ChinmayShringi opened the pull request:
## Description

This PR implements comprehensive syscall handling in the RawPOSIX dispatcher, focusing on security and memory safety. Each syscall implementation has been carefully compared against its NaCl counterpart to ensure consistent behavior while leveraging Rust's safety features.

Key improvements include:
- Enhanced memory validation using Rust's type system
- Proper protection flags for memory access (READ/WRITE)
- Consistent error handling with POSIX-compliant errno codes
- Resource management at cage layer

Reference NaCl implementation: [nacl_syscall_common.c](https://github.com/Lind-Project/native_client/blob/5f51b67716d15133ec41f8180ab779d527141a75/src/trusted/service_runtime/nacl_syscall_common.c#L801)

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- `cargo build` - Verifies compilation and type safety
- Manual verification against NaCl implementation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

qianxichen233 commented on src/RawPOSIX/src/safeposix/dispatcher.rs:
```diff
@@ -203,79 +203,197 @@ pub fn lind_syscall_api(
+            let length = arg2 as usize;
+            let cage = interface::cagetable_getref(cageid);
+        
+            // TODO(Security): Need to implement the following NaCl-style security checks:
```
These are already handled in `munmap_handler`

ChinmayShringi replied:
Okay, removed

qianxichen233 commented on src/RawPOSIX/src/safeposix/dispatcher.rs:
```diff
@@ -203,79 +203,197 @@ pub fn lind_syscall_api(
+                );
+            }
+        
+            // TODO(Security): Need NaClSysCommonMmapCheck equivalent to validate:
```
also already handled in `mmap_handler`

ChinmayShringi replied:
Okie, removed

qianxichen233 commented on src/RawPOSIX/src/safeposix/dispatcher.rs:
```diff
@@ -323,53 +457,75 @@ pub fn lind_syscall_api(
             
+            // Convert to reference for connect operation
             let remoteaddr = match Ok::<&interface::GenSockaddr, i32>(&addr) {
                 Ok(addr) => addr,
-                Err(_) => panic!("Failed to get sockaddr"), // Handle error appropriately
+                Err(_) => return syscall_error(Errno::EFAULT, "connect", "sockaddr conversion failed"),
             };
```
The code here looks weird to me. The `addr` is already explicitly constructed as `Ok` here, so I believe it should never go into `Err` branch?

ChinmayShringi replied:
Reverted

qianxichen233 commented on src/RawPOSIX/src/safeposix/dispatcher.rs:
```diff
@@ -323,53 +457,75 @@ pub fn lind_syscall_api(
+        
+            // Convert to reference for bind operation
             let localaddr = match Ok::<&interface::GenSockaddr, i32>(&addr) {
                 Ok(addr) => addr,
-                Err(_) => panic!("Failed to get sockaddr"), // Handle error appropriately
+                Err(_) => return syscall_error(Errno::EFAULT, "bind", "sockaddr conversion failed"),
             };
```
same question here

ChinmayShringi replied:
Reverted to old panic

qianxichen233 commented on src/RawPOSIX/src/safeposix/dispatcher.rs:
```diff
@@ -379,56 +535,104 @@ pub fn lind_syscall_api(
+            
+            // Get and validate fd sets
+            // NaCl: Uses NaClCopyInFromUser for fd_set validation
+            let readfds = match interface::get_fdset(arg2) {
```
readfds (or arg2) is a pointer that points to fdset struct. So I think we also need to do things like `check_and_convert_addr_ext` for it. Same for writefds, errorfds and timeout

ChinmayShringi replied:
Done

qianxichen233 commented on src/RawPOSIX/src/safeposix/dispatcher.rs:
```diff
@@ -510,101 +753,153 @@ pub fn lind_syscall_api(
-                Ok(addr) => interface::get_statdatastruct(addr).unwrap(),
+                Ok(addr) => match interface::get_statdatastruct(addr) {
+                    Ok(val) => val,
+                    Err(errno) => return syscall_error(Errno::EFAULT, "fxstat", "invalid stat data format"),
```
EFAULT typically refers to invalid memory address. But the error here means the struct of stat is not correct. So I think it might be better to find a more appropriate errno name. This may apply to other places since I saw EFAULT is hugely used here and I believe some of them are also not appropriate

rennergade replied:
I believe these are correct since the errors are triggered by pointer checking.

qianxichen233 replied:
I think this error is triggered by type conversion? The pointer checking function is `check_and_convert_addr_ext` if I understand correctly

rennergade replied:
that being said I don't think we have to change from the previous format? the internal function already returns syscall_error

qianxichen233 replied:
yeah the type conversion functions are already returning a syscall_error so I guess we can just return it

ChinmayShringi replied:
Reverted to previous format

rennergade commented on src/RawPOSIX/src/safeposix/dispatcher.rs:
```diff
@@ -203,79 +203,197 @@ pub fn lind_syscall_api(
 
     let ret = match call_number {
         WRITE_SYSCALL => {
+            // NaCl equivalent: NaClSysWrite
```
Lets get rid of all of these "NaCl equivalent comments" Its good that you checked all of these but we don't need to reference nacl in our new code.

ChinmayShringi replied:
Okie

rennergade commented on src/RawPOSIX/src/safeposix/dispatcher.rs:
```diff
@@ -203,79 +203,197 @@ pub fn lind_syscall_api(
             let fd = arg1 as i32;
-            let count = arg3 as usize;
+        
+            // Security: Clamp count to prevent integer overflow attacks
```
similarly i dont think we need this part anymore, we mostly just need to check the pointers and protections

ChinmayShringi replied:
Okie

rennergade commented on src/RawPOSIX/src/safeposix/dispatcher.rs:
```diff
@@ -203,79 +203,197 @@ pub fn lind_syscall_api(
+            }
+        
+            // Get cage reference for memory operations
+            // NaCl equivalent: struct NaClApp *nap = natp->nap
```
remove nacl reference

ChinmayShringi replied:
Okie, removed

rennergade reviewed (changes requested):
Looks very thorough. 

This is a very rare situation where I'll request to remove comments but let's get rid of all the references to NaCl in the code here. If the comment is just like a title its unescessary, if its referencing what nacl does, just change the comment to explaining the use instead of referencing nacl.

Yaxuan-w commented:
I just want to confirm if we’re on the same page. Does this PR focus on the dispatcher optimization for Wasmtime?

Yaxuan-w reviewed (commented):
The overall logic of the code looks good to me. However, I suggest updating the PR title/description to: 

> Optimize and enhance rawposix-wasmtime’s memory management by adopting the approach used in NaCl. 

(If I understand everything correctly)

rennergade commented:
> The overall logic of the code looks good to me. However, I suggest updating the PR title/description to:
> 
> > Optimize and enhance rawposix-wasmtime’s memory management by adopting the approach used in NaCl.
> 
> (If I understand everything correctly)

Yeah agree could be worded a bit better, something like "add memory validation  to RawPOSIX dispatcher" probably sums it up.

ChinmayShringi commented:
Hi,
This PR addresses issues with the `check_and_convert_addr_ext` function where `PROT_WRITE` and `PROT_READ` were incorrectly passed for some functions. To resolve this, I reviewed and compared all relevant syscalls in our implementation with the NaCl implementation, adding any missing checks and correcting the protection flags to ensure proper alignment. These changes enhance the robustness and correctness of our syscall behavior, ensuring consistency with NaCl's implementation. I have just added what I though was missing.

rennergade commented on src/RawPOSIX/src/interface/mem.rs:
```diff
@@ -21,8 +21,9 @@ pub fn round_up_page(length: u64) -> u64 {
-    let parent_base = parent_vmmap.base_address.unwrap();
-    let child_base = child_vmmap.base_address.unwrap();
+    // added _ to suppress the warning
+    let _parent_base = parent_vmmap.base_address.unwrap();
```
@qianxichen233 are these needed if theres a non use warning?

qianxichen233 replied:
yeah these are not needed anymore, we should delete them

rennergade commented on src/RawPOSIX/src/lib.rs:
```diff
@@ -1,4 +1,5 @@
-#![feature(lazy_cell)]
+// Rm for warning this line since lazy_cell is now stable
```
if we can remove it just remove it and not comment

rennergade commented on src/RawPOSIX/src/safeposix/dispatcher.rs:
```diff
@@ -203,79 +204,161 @@ pub fn lind_syscall_api(
-            let count = arg3 as usize;
+        
+            // Check bounds using MAX_IO_BUFFER_BYTES
+            let count = std::cmp::min(arg3 as usize, i32::MAX as usize);
```
I'm not sure if this step is necessary. @qianxichen233 thoughts?

qianxichen233 replied:
MAKE_SYSCALL does use u64 for all the argument, but the entry of `write` in glibc restricted the type to be `size_t`, which should be 32-bit. Unless we assume user has the ability to bypass glibc syscall entry and invoke MAKE_SYSCALL directly, I think this check is not necessary

rennergade commented on src/RawPOSIX/src/interface/mem.rs:
```diff
@@ -21,8 +21,9 @@ pub fn round_up_page(length: u64) -> u64 {
 pub fn fork_vmmap(parent_vmmap: &Vmmap, child_vmmap: &Vmmap) {
-    let parent_base = parent_vmmap.base_address.unwrap();
-    let child_base = child_vmmap.base_address.unwrap();
+    // added _ to suppress the warning
```
All of these comments saying why you removed something shouldnt be in the code. Just remove the things, if you want to explain why you did it just leave a comment in the PR.

ChinmayShringi replied:
Okie I will raise a seperate PR for warning and fixing the test cases

rennergade reviewed (approved):
I think this is good to go. Though I think we can remove the cmp line in write(), still want @qianxichen233 feedback there.
